### PR TITLE
CRW-579 add three new scripts for adding digests, building map of digests in registries, and introspecting containers (including scratch ones) to extract files

### PR DESCRIPTION
**Pull request by nickboldt**
CRW-579 add three new scripts for adding digests, building map of digests in registries, and introspecting containers (including scratch ones) to extract files

Change-Id: I2a9c3ef028fd5e094cb247dfc85e2db3895b69a8
Signed-off-by: nickboldt <nboldt@redhat.com>

### CODE REVIEW

**Comment by nickboldt** [comment]
To run this:

```
cd /tmp
git clone git@github.com:eclipse/che-operator.git && cd che-operator
hub checkout https://github.com/eclipse/che-operator/pull/202
cd olm

./addDigests.sh -s eclipse-che-preview-openshift/deploy/olm-catalog/eclipse-che-preview-openshift -n eclipse-che-preview-openshift -v 7.9.0 
```

Output will be in `generated` folder.

**Comment by nickboldt** [comment]
Comparing the generated file (left) with the source file (right):

image

Now adding the relatedImages:

image

**Comment by nickboldt** [comment]
Note also since digests are not yet supported in Che I'm leaving the script to create content in `generated/` folder instead of overwriting the actual CSV.

**Comment by nickboldt** [comment]
See also https://github.com/redhat-developer/codeready-workspaces-operator/pull/16 for downstream WIP

**Comment by tolusha** [comment]
@nickboldt 
Thank you. You did my job. I was going to work on this in the current sprint)

**Comment by nickboldt** [comment]
You'll have to thank @davidfestal ... He wrote the first iteration.

**Comment by nickboldt** [comment]
There are lots of bugs in CRW 2.1 you could investigate instead... ;)

File: olm/addDigests.sh [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#   Red Hat, Inc. - initial API and implementation
+
+SCRIPTS_DIR=$(cd "$(dirname "$0")"; pwd)
+BASE_DIR="$(pwd)"
+
```
Comment (tolusha) [inline olm/addDigests.sh]:
```suggestion
command -v podman >/dev/null 2>&1 || { echo "podman is not installed. Aborting."; exit 1; }
command -v yq >/dev/null 2>&1 || { echo "yq is not installed. Aborting."; exit 1; }
```

**Comment by tolusha** [comment]
@nickboldt 
It doesn't work for me
```bash
./addDigests.sh -s eclipse-che-preview-openshift/deploy/olm-catalog/eclipse-che-preview-openshift -n eclipse-che-preview-openshift -v 7.9.0 rm -rf generated 
[INFO] Get images from CSV /tmp/che-operator/olm/generated/eclipse-che-preview-openshift/7.9.0/eclipse-che-preview-openshift.v7.9.0.clusterserviceversion.yaml
>>>>>>>> start
[INFO] Pulling quay.io/eclipse/che-plugin-registry:7.9.0 ...
Trying to pull quay.io/eclipse/che-plugin-registry:7.9.0...
Getting image source signatures
Copying blob 30230d5fd633 done  
Copying blob 8eb9b1e99c64 done  
Copying blob e6a50b627bcb done  
Copying blob d67c4dfe3627 done  
Copying blob 1d9c8592f4f9 done  
Copying blob 50982eb4e7ff done  
Copying blob 58eb5d519e3a done  
Copying blob 8e15eac29301 done  
Copying blob 22d7ee08c96d done  
Copying blob 7cfff95d9858 done  
Copying blob 55d2b0950467 done  
Copying config ffc1dfd9d2 done  
Writing manifest to image destination
Storing signatures
ERRO[0024] Error while applying layer: ApplyLayer exit status 1 stdout:  stderr: there might not be enough IDs available in the namespace (requested 0:54 for /run/lock/lockdev): lchown /run/lock/lockdev: invalid argument 
  ApplyLayer exit status 1 stdout:  stderr: there might not be enough IDs available in the namespace (requested 0:54 for /run/lock/lockdev): lchown /run/lock/lockdev: invalid argument
Error: error pulling image "quay.io/eclipse/che-plugin-registry:7.9.0": unable to pull quay.io/eclipse/che-plugin-registry:7.9.0: unable to pull image: Error committing the finished image: error adding layer with blob "sha256:e6a50b627bcb03d96996bb8e836ecb178eae7425636e3424d9e8d33a918768dd": ApplyLayer exit status 1 stdout:  stderr: there might not be enough IDs available in the namespace (requested 0:54 for /run/lock/lockdev): lchown /run/lock/lockdev: invalid argument
```

**Comment by tolusha** [comment]
And in case of docker
```
[INFO] Pulling quay.io/eclipse/che-plugin-registry:7.9.0 ...
Error: No such container: quay.io-eclipse-che-plugin-registry-7.9.0-1584608804
[INFO] Extract from container ...
tar: Pattern matching characters used in file names
tar: Use --wildcards to enable pattern matching, or --no-wildcards to suppress this warning
tar: var/www/html/*/external_images.txt: Not found in archive
tar: Exiting with failure status due to previous errors
```

File: olm/dockerContainerExtract.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+if [[ ! $1 ]]; then
+  echo "Usage: $0 CONTAINER [tar-extraction-flags]"
+  echo "Usage: $0 quay.io/crw/operator-metadata:latest"
+  echo "Usage: $0 quay.io/crw/devfileregistry-rhel8:latest var/www/html/*/external_images.txt"
+  echo "Usage: $0 quay.io/crw/pluginregistry-rhel8:latest var/www/html/*/external_images.txt"
+  exit
+fi
+
+PODMAN=podman # or user docker
```
Comment (tolusha) [inline olm/dockerContainerExtract.sh]:
```suggestion
PODMAN=docker # or user podman
```

File: olm/dockerContainerExtract.sh [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+# create local container
+${PODMAN} rm -f "${tmpcontainer}" 2>&1 >/dev/null || true
+# use sh for regular containers or ls for scratch containers
+${PODMAN} create --name="${tmpcontainer}" $container sh 2>&1 >/dev/null || ${PODMAN} create --name="${tmpcontainer}" $container ls 2>&1 >/dev/null 
+
+# export and unpack
+${PODMAN} export "${tmpcontainer}" > /tmp/${tmpcontainer}.tar
+rm -fr "$unpackdir"; mkdir -p "$unpackdir"
+echo "[INFO] Extract from container ..."
+tar xf /tmp/${tmpcontainer}.tar -C "$unpackdir" $*
```
Comment (tolusha) [inline olm/dockerContainerExtract.sh]:
```suggestion
tar xf /tmp/${tmpcontainer}.tar --wildcards -C "$unpackdir" $*
```

**Comment by tolusha** [comment]
Updated scripts

**Pull request merged**